src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -20,17 +20,106 @@ internal sealed class SocketAsyncContext
     {
         private abstract class AsyncOperation
         {
+            private enum State
+            {
+                Waiting = 0,
+                Running = 1,
+                Complete = 2,
+                Cancelled = 3
+            }
+
+            private int _state; // Actually AsyncOperation.State
+
             public AsyncOperation Next;
+            protected object CallbackOrEvent;
             public SocketError ErrorCode;
             public byte[] SocketAddress;
             public int SocketAddressLen;
 
+            public ManualResetEventSlim Event { set { CallbackOrEvent = value; } }
+
             public AsyncOperation()
             {
+                _state = (int)State.Waiting;
                 Next = this;
             }
 
-            public abstract void Complete();
+            public void QueueCompletionCallback()
+            {
+                Debug.Assert(!(CallbackOrEvent is ManualResetEventSlim));
+                Debug.Assert(_state != (int)State.Cancelled);
+
+                ThreadPool.QueueUserWorkItem(o => ((AsyncOperation)o).DoCallback(), this);
```

Comment:
Consider adding a debug-only field and check to ensure that QueueCompletionCallback is invoked at most once.  I realize it would entail a #if DEBUG for the field, but we've done that in a few places where it's important... catching an error around multiple callbacks here would rate, I think.